src/python/pants/cache/artifact.py

```diff
@@ -109,30 +107,9 @@ def collect(self, paths):
         self._relpaths.add(relpath)
 
   def extract(self):
-    try:
-      with open_tar(self._tarfile, 'r', errorlevel=2) as tarin:
-        # Note: We create all needed paths proactively, even though extractall() can do this for us.
-        # This is because we may be called concurrently on multiple artifacts that share directories,
-        # and there will be a race condition inside extractall(): task T1 A) sees that a directory
-        # doesn't exist and B) tries to create it. But in the gap between A) and B) task T2 creates
-        # the same directory, so T1 throws "File exists" in B).
-        # This actually happened, and was very hard to debug.
-        # Creating the paths here up front allows us to squelch that "File exists" error.
-        paths = []
-        dirs = set()
-        for tarinfo in tarin.getmembers():
-          paths.append(tarinfo.name)
-          if tarinfo.isdir():
-            dirs.add(tarinfo.name)
-          else:
-            dirs.add(os.path.dirname(tarinfo.name))
-        for d in dirs:
-          try:
-            os.makedirs(os.path.join(self._artifact_root, d))
-          except OSError as e:
-            if e.errno != errno.EEXIST:
-              raise
-        tarin.extractall(self._artifact_root)
-        self._relpaths.update(paths)
-    except tarfile.ReadError as e:
-      raise ArtifactError(str(e))
+    # Note(yic): unlike the python implementation before, now we do not update self._relpath
+    # after the extraction.
+    result = Native.NATIVE_SINGLETON.decompress_tarball(self._tarfile.encode('utf-8'),
```

Comment:
Rather than setting up a singleton here, the `CacheFactory` is passed the `Task` which should have a `self.context._scheduler` that you could wire through.

Reply:
Thanks for the pointer. However, the plumbing may be a little cumbersome, will dig more.

Reply:
Moved to a class variable assigned at CacheFactory stage.